### PR TITLE
rpc: Support specifying HTTP client in RPC dialing

### PR DESCRIPTION
**Pull request by stevenroose**
Adds a minimal interface that captures `http.Client` and adds a new method `rpc.DialHTTPClient` that takes a client using that interface. The existing `rpc.DialHTTP` method is then alternatively implemented by using the new `rpc.DialHTTPClient` method provided with a standard `*http.Client`.

### CODE REVIEW

File: rpc/http.go [hunk 1]
```diff
@@ -40,8 +40,14 @@ const (
 
 var nullAddr, _ = net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 
+// HttpClient contaisn the methods of http.Client that we require.
```
Comment (ernestas2k) [inline rpc/http.go]: typo 'contaisn'

Comment (stevenroose) [reply rpc/http.go]: Haha, thanks. Fixed it.

**Comment by stevenroose** [comment]
ping..

File: rpc/http.go [hunk 2]
```diff
@@ -40,8 +40,14 @@ const (
 
 var nullAddr, _ = net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 
+// HttpClient contains the methods of http.Client that we require.
+// This allows users to provide wrapped HTTP clients for use with the RPC.
+type HttpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
```
Comment (karalabe) [inline rpc/http.go]: I don't think it's a good idea to define this interface here, as any eventual extra method (that's provided by Go's http.Client) will result in breaking the interface. Let's just directly require specifying a `*http.Client`. That way we're future proof.

File: rpc/http.go [hunk 3]
```diff
@@ -65,8 +71,9 @@ func (hc *httpConn) Close() error {
 	return nil
 }
 
-// DialHTTP creates a new RPC clients that connection to an RPC server over HTTP.
-func DialHTTP(endpoint string) (*Client, error) {
+// DialHTTPClient creates a new RPC clients that connection to an RPC server over HTTP
+// using the provided HTTP Client.
+func DialHTTPClient(endpoint string, client HttpClient) (*Client, error) {
```
Comment (karalabe) [inline rpc/http.go]: The usual "way" in Go code when initializing something with an extra parameter is `With`. I.e. `DialHTTPWithClient`. Please rename.

File: rpc/http.go (hunk 2 above)
Comment (stevenroose) [reply rpc/http.go]: Yeah had that first, but it makes it impossible to use wrappers of an http.Client like proxies or monitors.

File: rpc/http.go (hunk 3 above)
Comment (stevenroose) [reply rpc/http.go]: Yeah that makes sense.

File: rpc/http.go (hunk 2 above)
Comment (karalabe) [reply rpc/http.go]: AFAIK that's what [`http.Client.Transport`](https://golang.org/pkg/net/http/#Client) is for:

        // Transport specifies the mechanism by which individual
        // HTTP requests are made.
        // If nil, DefaultTransport is used.
        Transport RoundTripper

You can define all proxy and monitoring methods inside the http Client and that's how other libraries use it too (e.g. Google's Go APIs https://github.com/google/google-api-go-client/issues/146#issuecomment-228140570).

Comment (stevenroose) [reply rpc/http.go]: Hmm, didn't know about the Transport field. That's a smart thing they did there. Will change, one sec.

**Pull request merged**